### PR TITLE
Pattern acceleration support

### PR DESCRIPTION
**Pull request by elliVM**
## Pattern acceleration
- Replace fixed number of bloom filter tables with the support for dynamic number of tables
- Get bloom filter tables from SQL metadata
- Check if regex patterns stored in each bloom filter table matches  archive query search term tokenized values
- Left join only those bloom filter tables that had pattern match
- Filter query using bloommatch function for each logfile or check if filter is null on all joined tables
- Some refactoring on com.teragrep.pth_06.planner.walker and tests

### CODE REVIEW

File: database/bloomdb.sql [hunk 1]
```diff
@@ -1,6 +1,6 @@
 /*
- * This program handles user requests that require archive access.
- * Copyright (C) 2022  Suomen Kanuuna Oy
+ * Teragrep DPL to Catalyst Translator PTH-10
```
Comment (eemhu) [inline database/bloomdb.sql:2]: pth-06?

**Review by eemhu — COMMENTED** [review]
Some of the objects are not immutable, is it fixable or perhaps out of the scope of this PR?

File: database/bloomdb.sql (hunk 1 above)
Comment (elliVM) [reply database/bloomdb.sql]: fixed

**Comment by elliVM** [comment]
All new classes now have only final values and public methods always return the same value after instantiation

**Comment by elliVM** [comment]
All new classes now have only final values and public methods always return the same value after instantiation (weakly immutable). I included some refactoring of old classes but I think the project would require a larger refactoring to be more object-oriented. Made project issues regarding refactoring

**Review by eemhu — APPROVED** [review]
Changes look good to me!

File: src/main/java/com/teragrep/pth_06/planner/BloomFilterTempTable.java [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+
+import static com.teragrep.pth_06.jooq.generated.bloomdb.Bloomdb.BLOOMDB;
+import static org.jooq.impl.SQLDataType.*;
+
+/**
```
Comment (kortemik) [inline src/main/java/com/teragrep/pth_06/planner/BloomFilterTempTable.java:68]: Please write detailed javadoc about the logic behind the table structure and the query pattern. It is not very obvious to decode the schema and the query from the code.

Comment (elliVM) [reply src/main/java/com/teragrep/pth_06/planner/BloomFilterTempTable.java:67]: [39adb13](https://github.com/teragrep/pth_06/pull/65/commits/39adb13cd6a4f6b1820ab171ff4951c0affb99ec)  Added better javadoc for BloomFilterTempTable in an effort to clear up the logic.

File: src/main/java/com/teragrep/pth_06/planner/PatternMatch.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        final Condition finalPatternCondition = patternCondition;
+        List<Table<?>> tables = ctx.meta()
+                .filterSchemas(s -> s.equals(BLOOMDB))
+                .filterTables(t -> !t.equals(BLOOMDB.FILTERTYPE))
+                .filterTables(t -> ctx.select((Field<ULong>) t.field("id"))
+                        .from(t)
+                        .leftJoin(BLOOMDB.FILTERTYPE)
+                        .on(BLOOMDB.FILTERTYPE.ID.eq((Field<ULong>) t.field("filter_type_id")))
+                        .where(finalPatternCondition)
+                        .limit(1)
```
Comment (kortemik) [inline src/main/java/com/teragrep/pth_06/planner/PatternMatch.java:95]: why limit 1 is present? this limits the search token to make only one of the possible bloomfilters as a candidate instead of having all as possible candidates?

Comment (kortemik) [reply src/main/java/com/teragrep/pth_06/planner/PatternMatch.java:95]: please comment the code to indicate the purpose

File: src/main/java/com/teragrep/pth_06/planner/walker/conditions/ElementCondition.java [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+ * Creates a query condition from provided dom element
+ */
+public final class ElementCondition {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ElementCondition.class);
+
+    private final Element element;
+    private final ConditionConfig config;
+    private final Set<Table<?>> tableSet;
+    private final long bloomTermId;
+    private final List<Condition> conditionCache;
```
Comment (kortemik) [inline src/main/java/com/teragrep/pth_06/planner/walker/conditions/ElementCondition.java:71]: why a cache is being used?

File: src/main/java/com/teragrep/pth_06/planner/BloomFilterTempTable.java [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+ */
+public final class BloomFilterTempTable {
+    private static final Logger LOGGER = LoggerFactory.getLogger(BloomFilterTempTable.class);
+
+    private final DSLContext ctx;
+    private final Table<?> parentTable;
+    private final Table<Record> tableName;
+    private final long bloomTermId;
+    private final Set<Token> tokenSet;
+    private final List<Condition> cache;
```
Comment (kortemik) [inline src/main/java/com/teragrep/pth_06/planner/BloomFilterTempTable.java:101]: why a cache is being used?

File: src/test/java/com/teragrep/pth_06/walker/XmlWalkerTest.java [hunk 6]
```diff
@@ -46,230 +44,481 @@
...
+        Assertions.assertDoesNotThrow(() -> {
+            String q = "<index value=\"haproxy\" operation=\"NOT_EQUALS\"/>";
+            String e = "not (\"streamdb\".\"stream\".\"directory\" like 'haproxy')";
+            String result = streamWalker.fromString(q).toString();
+            Assertions.assertEquals(e, result);
+        });
     }
 
     @Test
-    void fromStringIntendTest() throws Exception {
```
Comment (kortemik) [inline src/test/java/com/teragrep/pth_06/walker/XmlWalkerTest.java:75]: changing existing test will not do, please revert the changes.

**Review by kortemik — CHANGES_REQUESTED** [review]

File: src/main/java/com/teragrep/pth_06/planner/PatternMatch.java (hunk 3 above)
Comment (elliVM) [reply src/main/java/com/teragrep/pth_06/planner/PatternMatch.java:95]: All tables in bloomdb are checked in turn by the filterTables() function, limit is just for the single table to avoid loading all of it's rows to memory, since it makes the filter if it has even just 1 matching row.

I will add the comment to clear the purpose

File: src/main/java/com/teragrep/pth_06/planner/BloomFilterTempTable.java (hunk 5 above)
Comment (elliVM) [reply src/main/java/com/teragrep/pth_06/planner/BloomFilterTempTable.java:101]: to limit the SQL operations if the Condition is already generated. Should I refactor them to work without caches?

**Comment by elliVM** [comment]
Refactoring of old code split into another PR

**Comment by elliVM** [comment]
Moving PR to another branch
